### PR TITLE
DEVX-156 add docs for state sync

### PR DESCRIPTION
**Pull request by lojzatran**
#### Summary

JIRA: https://commercetools.atlassian.net/browse/DEVX-156

### CODE REVIEW

File: docs/sdk2/usage/STATE_SYNC.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+#### Reference Resolution 
+
+`Transitions` are a way to describe possible transformations of the current state to other states of the same type (for example Initial -> Shipped). When performing a [SetTransitions](https://docs.commercetools.com/api/projects/states#set-transitions), an array of [ResourceIdentifiers](https://docs.commercetools.com/api/projects/states#ctp:api:type:StateResourceIdentifier) to State is needed.
+In commercetools, a [ResourceIdentifier](https://docs.commercetools.com/api/projects/states#ctp:api:type:StateResourceIdentifier) can be created by providing either the key or the ID. 
+When the key to a referenced state is provided with a `ResourceIdentifier`, the sync will resolve the resource with the given key and use the ID of the found resource to create or update a reference. 
+
+| ResourceIdentifier Field | Type                                       |
+|:-------------------------|:-------------------------------------------|
+| `transitions`            | Array of StateResourceIdentifiers to State |
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:75]:
```suggestion
| `transitions`            | Array of StateResourceIdentifiers | 
```

File: docs/sdk2/usage/STATE_SYNC.md [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+
+* sync exception
+* state draft from the source 
+* state of the target project (only provided if an existing state could be found)
+
+````java
+ final Logger logger = LoggerFactory.getLogger(StateSync.class);
+ final StateSyncOptions stateSyncOptions = StateSyncOptionsBuilder
+         .of(projectApiRoot)
+         .warningCallback((syncException, draft, state, updateActions) ->
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:164]:
```suggestion
         .warningCallback((syncException, draft, state) -> 
```

File: docs/sdk2/usage/STATE_SYNC.md [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+````java
+final TriFunction<
+        List<StateUpdateAction>, StateDraft, State, List<StateUpdateAction>> 
+            beforeUpdateStateCallback =
+            (updateActions, newStateDraft, oldState) ->  updateActions.stream()
+                    .filter(updateAction -> !(updateAction instanceof StateRemoveRolesAction))
+                    .collect(Collectors.toList());
+                        
+final StateSyncOptions stateSyncOptions = 
+        StateSyncOptionsBuilder.of(sphereClient).beforeUpdateCallback(beforeUpdateStateCallback).build();
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:186]:
```suggestion
        StateSyncOptionsBuilder.of(projectApiRoot).beforeUpdateCallback(beforeUpdateStateCallback).build();
```

File: docs/sdk2/usage/STATE_SYNC.md [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+Please refer to [example in product sync document](PRODUCT_SYNC.md#example-set-publish-stage-if-category-references-of-given-product-draft-exists).
+
+##### batchSize
+A number that could be used to set the batch size with which states are fetched and processed,
+as states are obtained from the target project on commercetools platform in batches for better performance. The algorithm accumulates up to `batchSize` resources from the input list, then fetches the corresponding states
+from the target project on the commecetools platform in a single request. Playing with this option can slightly improve or reduce processing speed. If it is not set, the default batch size is 50 for state sync.
+
+````java                         
+final StateSyncOptions stateSyncOptions = 
+         StateSyncOptionsBuilder.of(sphereClient).batchSize(30).build();
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:203]:
```suggestion
         StateSyncOptionsBuilder.of(projectApiRoot).batchSize(30).build();
```

File: docs/sdk2/usage/STATE_SYNC.md [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+```
+
+As soon, as the referenced StateDrafts are supplied to the sync, the draft will be created/updated and the 
+`CustomObject` will be removed from the target project.
+
+Keeping the old custom objects around forever can negatively influence the performance of your project and the time it takes to restore it from a backup. Deleting unused data ensures the best performance for your project. Please have a look into the [Cleanup guide](CLEANUP_GUIDE.md) to cleanup old unresolved custom objects.
+
+#### More examples of how to use the sync
+ 
+ 1. [Sync usages](https://github.com/commercetools/commercetools-sync-java/tree/master/src/integration-test/java/com/commercetools/sync/integration/externalsource/states/StateSyncIT.java).
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:293]:
```suggestion
 1. [Sync usages](#todo).
```

File: docs/sdk2/usage/STATE_SYNC.md [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+```
+
+### Build particular update action(s)
+
+Utility methods provided by the library to compare the specific fields of a `State` and a new `StateDraft`, and in turn builds
+ the update action. One example is the `buildSetNameAction` which compares names:
+````java
+Optional<StateUpdateAction> updateAction = StateUpdateActionUtils.buildSetNameAction(oldState, stateDraft);
+````
+More examples of those utils for different types can be found [here](https://github.com/commercetools/commercetools-sync-java/tree/master/src/test/java/com/commercetools/sync/states/utils/StateUpdateActionUtilsTest.java).
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:313]:
```suggestion
More examples of those utils for different types can be found [here](#todo).
```

File: docs/sdk2/usage/STATE_SYNC.md [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+
+
+## Migration Guide
+
+The cart-discount-sync uses the [JVM-SDK-V2](http://commercetools.github.io/commercetools-sdk-java-v2), therefore ensure you [Install JVM SDK](https://docs.commercetools.com/sdk/java-sdk-getting-started#install-the-java-sdk) module `commercetools-sdk-java-api` with
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:318]:
```suggestion
The state-sync uses the [JVM-SDK-V2](http://commercetools.github.io/commercetools-sdk-java-v2), therefore ensure you [Install JVM SDK](https://docs.commercetools.com/sdk/java-sdk-getting-started#install-the-java-sdk) module `commercetools-sdk-java-api` with
```

**Review by salander85 — CHANGES_REQUESTED** [review]

**Comment by lojzatran** [comment]
@salander85 please recheck: https://github.com/commercetools/commercetools-sync-java/pull/1044/commits/dd75153a10295ebb75e9e0f08315ec8a0845ab3f 🙇

File: docs/sdk2/usage/STATE_SYNC.md [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+            };
+```
+
+### Build StateDraft (syncing from external project)
+
+The state-sync expects a list of `StateDraft`s to process. If you use java-sync-library to sync your states from any external system into a commercetools platform project you have to convert your data into CTP compatible `StateDraft` type. This was done in previous version using `DraftBuilder`s.
+The V2 SDK do not have inheritance for `DraftBuilder` classes but the differences are minor and you can replace it easily. Here's an example:
+
+```java
+// StateDraftBuilder in v1 takes parameters 'name', 'cartPredicate', 'value', 'target', 'sortOrder', 'requiresDiscountCode'
```
Comment (salander85) [inline docs/sdk2/usage/STATE_SYNC.md:376]: Adjust the comment

**Review by salander85 — APPROVED** [review]
Please change the mentioned comment before merge.

File: docs/sdk2/usage/STATE_SYNC.md (hunk 8 above)
Comment (lojzatran) [reply docs/sdk2/usage/STATE_SYNC.md]: https://github.com/commercetools/commercetools-sync-java/pull/1044/commits/a3b6146ca828af2079f8991e138b8ce5124dce3e

**Pull request merged**